### ISSUE TITLE
Our upgrade/rollback process is incomplete

### ISSUE THREAD

**Issue by cygnusv** [nucypher/nucypher#1228]
Consider an upgrade of `StakingEscrow` that involves a change in a function signature, which implies an upgrade of `UserEscrowProxy` too.

In that case, the steps of the upgrade of `StakingEscrow` as follows:

1) Deploy new `StakingEscrow`
2) Upgrade target of `Dispatcher` for `StakingEscrow`
3) Deploy new `UserEscrowProxy`
4) Upgrade target of `UserEscrowLibraryLinker`

Currently, `StakingEscrowDeployer' covers steps 1 and 2 and stops right there. We have automation for steps 3 are 4 too in `UserEscrowProxyDeployer`, but it's not connected with the previous steps.

The question is: what should we do? @KPrasch and I think the solution is not to automate this more (i.e., connecting the first 2 steps with the last 2 steps in the same automatic workflow), but rather to improve human deployer awareness and obtain improved deployment status information.

**Comment by szotov** [comment]
I agree with manual upgrading (later can be done in one phase using some intermediary contract if we want). Also one suggestion: deploy first, upgrade later - change order of steps 2<->3

**Comment by cygnusv** [comment]
How do we know, in code, that an upgrade requires a new UserEscrowProxy?

**Comment by cygnusv** [comment]
Is it when the contract ABI changes?

**Comment by szotov** [comment]
I'm not sure that is possible automatically in code to do this, decision will have to be made manually. Logically upgrade could be only for one contract or bunch of them - when network will work only with all of upgrades.  Possible options:
1. Change only logic in one contract - deploy + upgrade this contract
2.  Change signatures for read-only (call) functions or other functions that not in UserEscrowProxy -  deploy + upgrade this contract
3. Change in other functions - deploy contract and UserEscrowProxy and then call bunch of upgrades
4. Same options (1-3) for multiple contracts in one upgrade - deploy all contracts (+ UserEscrowProxy if neccessary), then call upgrade for all dispatchers and linker

**Comment by cygnusv** [comment]
That's gold, Sergey! Gold!
